src/index.tsx

```diff
@@ -129,45 +127,58 @@ export function useSpeechRecognition(options: SpeechRecognitionOptions = default
   );
 
   const onRecognitionDisconnect = useCallback(() => {
-    dispatch(setStatus(SpeechRecognitionStatus.STOPPED));
     if (pauseAfterDisconnect) {
-      dispatch(setListening(false));
+      dispatch(setStatus(SpeechRecognitionStatus.STOPPED));
     } else if (recognition) {
       if (recognition.continuous) {
         startListening();
       } else {
-        dispatch(setListening(false));
+        dispatch(setStatus(SpeechRecognitionStatus.STOPPED));
       }
     }
     dispatch(setPauseAfterDisconnect(false));
   }, [pauseAfterDisconnect, recognition, startListening]);
 
   const onRecognitionError = useCallback(({ error, message }) => {
-    console.log("Speech recognition error detected: " + error);
-    console.log("Additional information: " + message);
+    dispatch(setStatus(SpeechRecognitionStatus.ERROR));
+    dispatch(setError(`${error}: ${message}`));
   }, []);
 
   useEffect(() => {
-    if (recognition && !listening) {
+    if (recognition) {
       recognition.continuous = options.continuous !== false;
       recognition.interimResults = options.interimResults;
       recognition.onresult = updateTranscript;
       recognition.onend = onRecognitionDisconnect;
       recognition.onerror = onRecognitionError;
     }
+  }, [
+    onRecognitionDisconnect,
+    onRecognitionError,
+    updateTranscript,
+    recognition,
+    options.continuous,
+    options.interimResults,
+  ]);
 
+  useEffect(() => {
     if (recognition && options && options.autoStart) {
-      recognition.start();
-      dispatch(setListening(true));
+      startListening();
     }
-  }, [listening, onRecognitionDisconnect, onRecognitionError, options, recognition, updateTranscript]);
+
+    return () => {
+      if (status === SpeechRecognitionStatus.STARTED) {
```

Comment:
should stop only be performed on clean up if autostart?

Reply:
I'd say that we want it cleaned up regardless of how it got there.